Update type-only import semantics to allow type queries

andrewbranch opened the pull request:
In #35200, a type-only import or export created a synthetic symbol which reflected only the type-side meaning of its target. This meant that the imported names were unusable in any value position. This PR removes the intermediate synthetic symbol and instead issues an error if the imported names appear in any non-ambient expression context. Namely, this allows for:

```ts
import type { BaseClass } from './base';

// 1. Type queries
let ConstructorType: typeof BaseClass;

// 2. Ambient contexts that require value meanings
declare class SubClass extends BaseClass {}
```

---

One oddity that arises with this approach is how to treat a namespace import that includes type-only exports:

```ts
// a.ts
class A {}
export type { A }

// b.ts
import * as a from './a';
console.log(a); // Empty object, because the export type was erased

type T = keyof typeof a; // `never` - because `typeof a` must be an empty object type

let a: a.A; // This is ok
let A: typeof a.A; // This doesn’t work because `typeof a` is an empty object :(
```

Essentially, there is a limitation that if you need to access the _value_ meaning of a type-only _export_, you can’t do so through a namespace import—you have to use a named (or default) import.

The distinction is subtle and confusing, but I think internally consistent, and unlikely to come up much in real code. At least, I think it’s better to start with this limitation than to be overly permissive in a way that could make the type system misrepresent the reality of the shape of a module.

---

Fixes #36040
Fixes #36003 (the part I said was a bug)
Fixes #36004

andrewbranch commented on src/compiler/checker.ts:
```diff
@@ -2587,6 +2532,15 @@ namespace ts {
+            if (!links.constEnumReferenced) {
+                links.constEnumReferenced = true;
+            }
+        }
```
This was surprisingly the ugliest part of the changes. Aliases that resolve to const enums are explicitly excluded from being marked as `symbol.referenced` because that’s what the emit resolver uses to determine if the import can be elided, and since const enums are inlined, the alias is, in a sense, not referenced. Meanwhile, the `symbol.isReferenced` property, which is used to report unused variables, is incapable of tracking whether an enum was referenced as a type or as a value, because name resolution of dotted names like `Enum.Member` looks for the meaning `SymbolFlags.Namespace`, which overlaps with both `SymbolFlags.Type` and `SymbolFlags.Value`. The distinction is necessary for `--importsNotUsedAsValues=error`, unless we want to treat const enums in some special way under that flag.

Does this make sense, or did I miss an easier/better way to do this?

(I wonder if const enums should be marked as `referenced` during checking, and the logic to treat them as unreferenced be deferred to the emit resolver?)

weswigham replied:
Proposal: Treat type only and normal imports the same in the checker, this way usage diagnostics are reported, but then in the transformer, always elide `import type` and `export type` statements in the TS transform (without needing to refer to the checker via the emit resolver).

andrewbranch replied:
> this way usage diagnostics are reported

They already are.

> but then in the transformer, always elide `import type` and `export type` statements in the TS transform

This is already how it works, both in #35200 and here. What’s new here is that since we retain the value meaning of aliases, the emit resolver started returning `true` for `isValueAliasDeclaration` and `isReferencedAliasDeclaration` if those aliases are used in expressions (which is an error). The effect was kind of a weird emit. I wasn’t sure if it was worth bothering with since the code had errors, but @DanielRosenwasser suggested so:

![image](https://user-images.githubusercontent.com/3277153/72103494-2dc16900-32de-11ea-9d4d-20e9cccf0cec.png)

However, this is a little beside the point of my question here—I’m talking about the `symbol.referenced` of const enums, which has been a long-standing behavior before any of my type-only work.

weswigham replied:
Right - my point is that if you ignore the emit resolver entirely when printing `import type| and `export type` and always the behavior of `symbol.referenced` no longer should matter, since you wouldn't be referencing it in the import/export type case. The references to those things, should they exist in value space, are fine to have the emit that expects the import be present, imo - the code's nonsensical anyway.

But anyways, re: the distinction - I'm pretty sure I remember at some point in the past Nathan had swapped `symbol.isReferenced` from a Boolean to `SymbolFags` encoding how it was referenced for a PR at some point. Doing that here could work well.

andrewbranch replied:
If the nonsensical `a_1.A` emit reference is ok, then yes, I can make a couple simplifications that might even help the perf situation. But as far as the const enums go, from my top comment:

> the symbol.isReferenced property, which is used to report unused variables, is incapable of tracking whether an enum was referenced as a type or as a value, because name resolution of dotted names like `Enum.Member` looks for the meaning `SymbolFlags.Namespace`, which overlaps with both `SymbolFlags.Type` and `SymbolFlags.Value`. 

So if a const enum is referenced in a type position, it _appears_ as if it was also referenced in a value position, as far as `isReferenced` goes. I could probably try to tighten up that namespace lookup.

andrewbranch replied:
> Doing that here could work well.

Oh, I may have misunderstood your suggestion. I thought you were suggesting to use `isReferenced`, but I think you may have been talking about changing `referenced` to flags.

andrewbranch commented:
@typescript-bot perf test this

andrewbranch commented:
Check time appears up ~2–4%. I think [this discussion](https://github.com/microsoft/TypeScript/pull/36092#discussion_r364952936) will have some bearing on what we can do, so I’ll hold off on doing anything specifically for perf until we make some other decisions.

andrewbranch commented:
@typescript-bot perf test this

andrewbranch commented:
@typescript-bot perf test this

weswigham commented on src/compiler/checker.ts:
```diff
@@ -1851,6 +1851,21 @@ namespace ts {
                         error(errorLocation, Diagnostics.Initializer_of_parameter_0_cannot_reference_identifier_1_declared_after_it, declarationNameToString(associatedDeclarationForContainingInitializer.name), declarationNameToString(<Identifier>errorLocation));
                     }
                 }
+                if (result && errorLocation && meaning & SymbolFlags.Value && !(errorLocation.flags & NodeFlags.Ambient) && isInExpressionContext(errorLocation)) {
```
Just a reminder from the design meeting, since it looks like the new reference cache isn't the perf issue, I'd wager it's the `isInExpressionContext` call here. It might be better to annotate the node with a flag corresponding context flag to this boolean during parse if that's the case, just like how we do for `NodeFlags.JavaScriptFile` and `NodeFlags.JsonFile`, and `NodeFlags.Ambient`.

andrewbranch replied:
I did some profiling after the meeting and it _seems_ like this is not problematic (`isInExpressionContext` had 1 ms aggregated total time in a ~5 second compilation), with the caveat that I haven’t yet been able to reliably reproduce as significant a perf difference as we’re seeing here. However, a node flag still sounds like a good idea (I was surprised one didn’t exist in the first place).

weswigham commented on src/compiler/checker.ts:
```diff
@@ -2552,12 +2474,33 @@ namespace ts {
             if (target) {
                 const markAlias = target === unknownSymbol ||
-                    ((target.flags & SymbolFlags.Value) && !isConstEnumOrConstEnumOnlyModule(target));
+                    ((target.flags & SymbolFlags.Value) && !isConstEnumOrConstEnumOnlyModule(target) && !getTypeOnlyAliasDeclaration(symbol));
```
If you wanted to bench the removal of the calls to `getTypeOnlyAliasDeclaration`, you'd also need to `undefined`-out here, too~

weswigham commented on src/compiler/checker.ts:
```diff
@@ -22858,7 +22802,7 @@ namespace ts {
 
                 // If react symbol is alias, mark it as refereced
-                if (reactSym.flags & SymbolFlags.Alias) {
+                if (reactSym.flags & SymbolFlags.Alias && !getTypeOnlyAliasDeclaration(reactSym)) {
```
And here~

weswigham commented on src/compiler/checker.ts:
```diff
@@ -29653,7 +29597,8 @@ namespace ts {
                 && rootSymbol.flags & SymbolFlags.Alias
                 && symbolIsValue(rootSymbol)
-                && !isConstEnumOrConstEnumOnlyModule(resolveAlias(rootSymbol))) {
+                && !isConstEnumOrConstEnumOnlyModule(resolveAlias(rootSymbol))
```
And here~

weswigham commented on src/compiler/checker.ts:
```diff
@@ -29653,7 +29597,8 @@ namespace ts {
                 && symbolIsValue(rootSymbol)
-                && !isConstEnumOrConstEnumOnlyModule(resolveAlias(rootSymbol))) {
+                && !isConstEnumOrConstEnumOnlyModule(resolveAlias(rootSymbol))
+                && !getTypeOnlyAliasDeclaration(rootSymbol)) {
```
And here~

weswigham commented on src/compiler/checker.ts:
```diff
@@ -34555,11 +34499,13 @@ namespace ts {
                 case SyntaxKind.ExportSpecifier:
-                    return isAliasResolvedToValue(getSymbolOfNode(node) || unknownSymbol);
+                    const symbol = getSymbolOfNode(node) || unknownSymbol;
+                    return isAliasResolvedToValue(symbol) && !getTypeOnlyAliasDeclaration(symbol);
```
And here~ (wow, this got added to more places than I thought)

weswigham commented on src/compiler/checker.ts:
```diff
@@ -34465,7 +34409,7 @@ namespace ts {
                 // We should only get the declaration of an alias if there isn't a local value
                 // declaration for the symbol
-                if (isNonLocalAlias(symbol, /*excludes*/ SymbolFlags.Value)) {
+                if (isNonLocalAlias(symbol, /*excludes*/ SymbolFlags.Value) && !getTypeOnlyAliasDeclaration(symbol)) {
```
And here~

andrewbranch commented:
@typescript-bot perf test this, and don’t give me bad news

andrewbranch commented:
@typescript-bot perf test this

andrewbranch commented:
@typescript-bot perf test this

DanielRosenwasser commented:
@typescript-bot pack this

DanielRosenwasser commented:
Hey, is this use-case covered?

```ts
import type { onInit } from "./framework-hooks";

interface Component {
  [onInit]?(): void;
}
```

DanielRosenwasser commented:
Yeah, seems like that is another ambient place where values have to be resolved.

Also worth testing in properties on `declare class`

```ts
import type { onInit } from "./framework-hooks";

declare class ConcreteComponent {
  [onInit](): void;
}
```

andrewbranch commented:
For a moment I thought we could also allow type-only imports in class property declarations without initializers:

```ts
class C {
  [onInit]: any;
}
```

But that’s actually a problem for ES class field semantics if I understand correctly. `declare` property declarations should Just Work™ because the ambient node flag will propagate down.

It’s also maybe worth noting that at this point, it would be very easy to allow people to use const enums as type-only imports (#36003), even though I think it’s pretty unintuitive.

DanielRosenwasser commented:
Yeah, that's definitely an error. Here's another fun case though: `abstract` members!

```ts
import type { onInit } from "./framework-hooks";

abstract class Component {
    abstract [onInit](): void;
}
```

andrewbranch commented:
@DanielRosenwasser I’ve just updated to include those cases, and switched to a NodeFlag-based approach as @weswigham suggested earlier since the list of syntactic forms to check for in a hot path was growing.

andrewbranch commented:
@typescript-bot perf-test this

DanielRosenwasser commented on src/compiler/types.ts:
```diff
@@ -574,15 +574,15 @@ namespace ts {
         JsonFile                                      = 1 << 25, // If node was parsed in a Json
         /* @internal */ TypeCached                    = 1 << 26, // If a type was cached for node at any point
-        /* @internal */ InNonEmittingNode             = 1 << 27, // If a node was in a context where expressions are never emitted: type queries, abstract members, computed property names of types.
+        /* @internal */ InNonEmittingExpression       = 1 << 27, // If a node was in a context where expressions are never emitted: type queries, abstract members, computed property names of types.
```
Can you tell me more about this? I'm not as familiar with this part of the codebase, and I'm not sure why we're changing it from Nodes to Expressions

andrewbranch replied:
I wrote that flag as part of this PR, just changing the meaning between commits. But I might have to do something different, because apparently it is still super hard to tell whether we’re in a type position or not unless we’re in the parser. 🙄 

The problem with setting context flags is that they limit what the incremental parser will reuse, so I was hesitant to set this across large type declaration like interfaces and object literal types. We also don’t have a ton of test coverage for incremental parsing so it’s hard to tell how much impact that would have on a typical source file. So I’m now looking to see if I can keep setting this flag on all type nodes, but wire in some help for the incremental parser here and there. I’ve done a similar optimization with ambient nodes in the past, so I’m hopeful: https://github.com/microsoft/TypeScript/pull/32849

DanielRosenwasser commented on src/compiler/utilities.ts:
```diff
@@ -1779,6 +1771,31 @@ namespace ts {
     }
 
+    export function isPartOfTypeQuery(node: Node) {
+        while (node.kind === SyntaxKind.QualifiedName || node.kind === SyntaxKind.Identifier) {
```
As a minor optimization, I think that `QualifiedName`s are always parts of types, but don't take my word for it maybe.

andrewbranch replied:
I think you’re right.

DanielRosenwasser replied:
This would be true if it weren't for `import foo = a.b.c`! There's `PropertyAccessEntityName` but I'm not sure if that helps.

andrewbranch replied:
But, this util currently has a reasonable name and a reasonable chance of reuse. Not sure an optimization here is worth the hit to the understandability of what we’re checking.

DanielRosenwasser commented on src/compiler/utilities.ts:
```diff
@@ -1779,6 +1771,31 @@ namespace ts {
+        return containerKind === SyntaxKind.InterfaceDeclaration || containerKind === SyntaxKind.TypeLiteral;
+    }
+
+    export function isAbstractDeclarationName(node: Node) {
```
Is this used anywhere?

DanielRosenwasser commented:
Looks like there's diagnostic conflicts

andrewbranch commented:
@typescript-bot perf test this again for good measure please

xiaoyu-tamu commented:
Is there any settings could disable this?? 

It's so annoying when use DI framework like nest.js.  I want it always import value not type.

andrewbranch commented:
It doesn’t do this unless you set `importsNotUsedAsValues: "error"` or `"preserve"` in your tsconfig.
